routing/router.go

```diff
@@ -1675,3 +1676,119 @@ func receiverAmtForwardPass(runningAmt lnwire.MilliSatoshi,
 
 	return runningAmt, nil
 }
+
+// outgoingFromIncoming computes the outgoing amount based on the incoming
+// amount by subtracting fees from the incoming amount. Note that this is not
+// exactly the inverse of incomingFromOutgoing, because of some rounding.
+func outgoingFromIncoming(incomingAmt lnwire.MilliSatoshi,
+	incoming, outgoing *unifiedEdge) lnwire.MilliSatoshi {
+
+	// Convert all quantities to floats to be able to hande negative values.
+	// The formulas to compute the outgoing amount involve terms with
+	// PPM*PPM*A, which can easily overflow an int64. Other terms may
+	// overflow as well, which is why we chose to work with float64 here.
+	A := float64(incomingAmt)
+	Ro := float64(outgoing.policy.FeeProportionalMillionths)
+	Bo := float64(outgoing.policy.FeeBaseMSat)
+	Ri := float64(incoming.inboundFees.Rate)
+	Bi := float64(incoming.inboundFees.Base)
+	PPM := float64(1_000_000)
+
+	// The following discussion was contributed by user feelancer21, see
+	//nolint:lll
+	// https://github.com/feelancer21/lnd/commit/f6f05fa930985aac0d27c3f6681aada1b599162a.
+
+	// The incoming amount Ai based on the outgoing amount Ao is computed by
+	// Ai = max(Ai(Ao), Ao), which caps the incoming amount such that the
+	// total node fee (Ai - Ao) is non-negative. This is commonly enforced
+	// by routing nodes.
+
+	// The function Ai(Ao) is given by:
+	// Ai(Ao) = (Ao + Bo + Ro/PPM) + (Bi + (Ao + Ro/PPM + Bo)*Ri/PPM), where
+	// the first term is the net amount (the outgoing amount plus the
+	// outbound fee), and the second is the inbound fee computed based on
+	// the net amount.
+
+	// Ai(Ao) can potentially become more negative in amplitude than Ao,
+	// which is why the above mentioned capping is needed. We can abbreviate
+	// Ai(Ao) with Ai(Ao) = m*Ao + n, where m and n are:
+	// m := (1 + Ro/PPM) * (1 + Ri/PPM)
```

Comment:
not sure I follow, could you explain that?

Reply:
Hm, I'm still a bit confused by the statement, since I'd rather say it's accurate in real space (up to a cast to compare to the int result) but not in integer space as you need to do an integer division twice with the accompanied rounding, but good it's resolved

Reply:
I thought the reason we add the 1 in the first place is because of the truncation of the remainder in integer space. By keeping it in real space I do not believe we need the `1 +` part of the expression. Maybe I misunderstand.

Reply:
It's not about rounding here, it's the exact formula, but factored to see how the fee rates compare to 1, to see if it results in a zero or negative overall term, see the og derivation https://github.com/feelancer21/lnd/blob/f6f05fa930985aac0d27c3f6681aada1b599162a/prototype/doc.md#outgoing---incoming.